### ISSUE TITLE
dinit readiness notification support

### ISSUE THREAD

**Issue by q66** [davmac314/dinit#67]
I was wondering if you'd be interested in having support for readiness notification of dinit itself, in a similar manner to how it handles it for services themselves, I was thinking something like having a launch parameter to pass a file descriptor number to which dinit would write after finishing startup of its initial services (`boot` or the ones passed on command line).

My use case for this is for `dinit-userservd`, which is Chimera's daemon that automatically manages user instances of `dinit` while keeping track of logins/logouts (via a PAM module; the PAM module connects to the daemon's socket, which is considered a login, and disconnection is considered a logout).

Right now I have a bit clunky mechanism in place - the daemon generates the `boot` service it starts `dinit` with, which then does startup of user services (your custom `boot.d` which lives in your home directory), and the daemon itself basically creates a named pipe, while the generated `boot` service pokes the write end of that pipe, which then signals the daemon and the daemon allows the login to proceed on PAM side.

If I was able to specify the file descriptor and the readiness was handled in `dinit` itself, I would be able to create a local unnamed pipe in-process and pass the write end to `dinit` (since it's a child process) and otherwise keep things the same (just without having to touch the filesystem).

I was going to implement this, but I first wanted to know your opinion and/or if you have any better ideas.

**Comment by davmac314** [comment]
Yes, I agree in principle that the feature would be a good one to have (and this use-case is worth supporting).

I don't really have any other ideas. One question that comes up is how to decide when `dinit` startup is complete - probably it should be tied to a single service, but should it be specified in the service description for that service or should it be specified via the command line (eg specify a named service, or perhaps just take the first service already specified defaulting to `boot` as usual). The only thing I would be against is making the service named "boot" always be special in this regard (as in, the name "boot" is special because it is the default service to start, but I would prefer it not be special in any other way; I hope that makes sense).

**Comment by q66** [comment]
I would simply wait for startup of the whole initial set (so `boot` by default, or whatever other services the user chooses to start) as that IMO makes the most sense (at least for this use case) but maybe we could support both (default to the given set, or if another specific service is provided, use that)

hardcoding `boot` sounds wrong to me as well, so i definitely would not do that

**Comment by davmac314** [comment]
> I would simply wait for startup of the whole initial set

That sounds ideal, really.

**Comment by q66** [comment]
alright, only need to work out the implementation details then (one could possibly create a special dummy service that has dependency links to whatever the user chooses to start and have it do the file descriptor poking when all that is met, but I'm thinking it would probably be better to do this at a lower level without involving that and use the event loop directly)

**Comment by q66** [comment]
well, the goal is explicitly "do not proceed with login until enabled user services have started (or timeout has been reached)" as othrewise it would be pointless to even check this (user services start necessary stuff like the d-bus session bus, which has to be up by the time the user is allowed to do anything)

**Comment by q66** [comment]
btw, i take it you have plans to stabilize the dinitctl protocol then? would be neat (possibly with a library C API to make talking to it more convenient?)

**Comment by q66** [comment]
the more I'm thinking about it, your solution would actually be pretty good - one could make it even more convenient by having dinit write the path to the ready socket on the file descriptor, so you could poll on the read end of the pipe, receive the path, then directly connect to it or pass it as a part of the call to dinitctl, then wait for that to finish...

**Comment by q66** [comment]
C++ API is okay too (i usually deal with C++ anyway so either works), i mean just having something that is available as a public library API (raw sockets can be a pain to deal with)
